### PR TITLE
Instant Upload now offers selective upload of Videos and Images.

### PR DESCRIPTION
**Pull request by Jon-Schneider**
This will close issue #762.

This is based on the branch 'ready_to_ios10', which looks like it's about to be merged. After 'ready_to_ios10 ' is merged I'll rebase this branch onto master.

BUGS & IMPROVEMENTS
- [X] Photos' switch enables videos' switch https://github.com/owncloud/ios/pull/782#issuecomment-251312484 [FIXED]
- [X] Files taken before option is enabled are being uploaded https://github.com/owncloud/ios/pull/782#issuecomment-251320996 [FIXED]
- [X] Background mode switchs itself when the current account is switched https://github.com/owncloud/ios/pull/782#issuecomment-251324649 [FIXED]
- [X] Files taken before enabling option are being uploaded after killing and opening the app https://github.com/owncloud/ios/pull/782#issuecomment-252221403  [FIXED]
- [X] Timestamps mixed https://github.com/owncloud/ios/pull/782#issuecomment-253736636 [FIXED]

### CODE REVIEW

**Comment by Jon-Schneider** [comment]
@jesmrec I would re-check if repeat photo uploading (Issue #755) is still occurring. 

I'm looking into the other Instant Upload bug reported on this Issue as well.

**Comment by Jon-Schneider** [comment]
The Instant Upload issue with non-active user's Background Instant Upload preference not being updated is odd because the DB update line is **'correctQuery = [db executeUpdate:@"UPDATE users SET background_instant_upload=? ", [NSNumber numberWithBool:newValue]];'**. Per the SQLite docs (https://www.sqlite.org/lang_update.html) **"If the UPDATE statement does not have a WHERE clause, all rows in the table are modified by the UPDATE"**, and in fact an identical path (with just a different row name) works to update all users preferences for instant upload. I'm not sure why there is a disparity - can anyone more experienced with SQLite care to weigh in? The offending file is **ManageAppSettingsDB.m**

**Comment by nasli** [comment]
@Jon-Schneider, this query is OK, right now the instant uploads is per all app not per account, you must update all the users after any instant upload change.

**Comment by Jon-Schneider** [comment]
@nasli Rebased onto master.

**Comment by Jon-Schneider** [comment]
@nasli That SQL will update the background instant upload preference for all users. Is there another place I need to update it? My understanding was that the active user is pulled from the store when changed, so there's no persistent objects that represent other users in memory that need updated.

**Comment by nasli** [comment]
@Jon-Schneider Related to this bug https://github.com/owncloud/ios/issues/755#issuecomment-247582434, the problem seem to be that this  "background instant upload" option is not updated to the new users which are added after the setting is already enabled. 

The persistence is only in the DB and for the alive user. This is what you should update, and in the case new users, check this parameter with the alive user, so this way the new user will be store with the same option.

**Comment by Jon-Schneider** [comment]
@nasli Fixed.

**Comment by nasli** [comment]
@jesmrec take into account for test plan:
-upgrade version
-enable/disabled different setting options

**Comment by jesmrec** [comment]
### Bug

Steps:

1. initially, both instant uploads disabled
2. Enable "Instant Upload Photos"

Actual Behaviour:

Instant Upload Photos -> Disabled
Instant Upload Videos -> Enabled

Expected Behaviour:

Instant Upload Photos -> Enabled
Instant Upload Videos -> Disabled

App settings

screen shot 2016-10-04 at 09 15 35

Tested with iPhone 9.3.5

**Comment by jesmrec** [comment]
### Bug

Steps:

1. initially, both instant uploads disabled
2. Enable "Instant Upload Photos"
3. Take photos and videos (only photos are uploaded -> correct)
4. Enable "Instant Upload Videos"

Actual Behaviour:

Videos taken before enabling "Instant Upload Videos" are uploaded

Expected Behaviour:

Videos taken before enabling "Instant Upload Videos" are not uploaded. Only have to be uploaded those videos taken after enabling "Instant Upload Videos"

iPhone 9.3.2

**Comment by jesmrec** [comment]
### Bug

Steps:

1. Add an account and enable instant upload for videos of photos
2. Enable Background
3. Add some accounts and switch to them (background stays enabled)
4. Disable Background
5. Switch account

Actual Behaviour:

Background becomes enabled

Expected Behaviour:

Background stays disabled. Background is a global variable for all the accounts. Only the user can switch it.

iPhone 9.3.2

cc @nasli @Jon-Schneider

**Comment by Jon-Schneider** [comment]
@jesmrec The first bug is fixed.

**Comment by jesmrec** [comment]
@Jon-Schneider fixed!!

**Comment by rperezb** [comment]
@hodyroff @pmaier1 working on this one too!

**Comment by Jon-Schneider** [comment]
@jesmrec The Background Instant Upload switch is fixed.

**Comment by Jon-Schneider** [comment]
@jesmrec Fixed the final bug. All 3 should be fixed.

**Comment by jesmrec** [comment]
This bug is similar to the last one but with a different step:

1. Set the following set up
   Instant Uploads Photos -> enabled
   Instant Uploads Videos -> disabled
(or viceversa)
2. Take Photos and Videos -> only Photos are uploaded 
3. Enable Instant Uploads Videos -> Videos are not uploaded (bug fixed!)
4. Kill the app and open it again

Actual Behaviour

Videos that were taken before enabling instant uploads are now uploaded

Expected behaviour

Videos not uploaded

Checked iPhone6 9.3.5, iPadAir 10.0.2

cc @Jon-Schneider @nasli

**Comment by jesmrec** [comment]
Perhaps we can take advantage of this feature. We have this bug from last version:

https://github.com/owncloud/ios/issues/755#issuecomment-247583486

also related with instant uploads

What do you mean @Jon-Schneider? is it fixable within this issue?

**Comment by Jon-Schneider** [comment]
Good catch @jesmrec, I was only updating the last upload date in memory, not in the database too.

**Comment by jesmrec** [comment]
@Jon-Schneider fixed! great job :)

**Comment by nasli** [comment]
Since this feature is done, let's merge this and continue fixing regression instant uploads in new branch.

@Jon-Schneider could you rebase with master, thanks!

**Comment by Jon-Schneider** [comment]
Rebased, can someone rerun the CI?

**Comment by Jon-Schneider** [comment]
@nasli Rebased and ready to merge.

File: Owncloud iOs Client/InstantUpload/InstantUpload.m [hunk 1]
```diff
@@ -61,30 +61,71 @@ - (void) dealloc {
...
-                if ([PHPhotoLibrary authorizationStatus] == PHAuthorizationStatusAuthorized) {
-                    ACTIVE_USER.instantUpload = YES;
-                    [ManageAppSettingsDB updateInstantUploadTo:YES];
-                    ACTIVE_USER.timestampInstantUpload = [[NSDate date] timeIntervalSince1970];;
+        if (enabled != self.videoInstantUploadEnabled) {
+            if (!self.videoInstantUploadEnabled) {
+                if (self.imageInstantUploadEnabled) {
+                    ACTIVE_USER.videoInstantUpload = YES;
+                    [ManageAppSettingsDB updateVideoInstantUploadTo:YES];
+                    ACTIVE_USER.timestampInstantUpload = [[NSDate date] timeIntervalSince1970];
```
Comment (nasli) [inline Owncloud iOs Client/InstantUpload/InstantUpload.m]: You can not use same time stamp for photos and videos, they are going to have different time stamps. If you enabled photos, take some pictures, after a while enabled instant videos, previous photos are not going to be uploaded. Review this

File: Owncloud iOs Client/DataBase/Queries/ManageUsersDB.m [hunk 2]
```diff
@@ -233,7 +235,7 @@ + (UserDto *) getUserByIdUser:(NSInteger) idUser {
     FMDatabaseQueue *queue = Managers.sharedDatabase;
     
     [queue inDatabase:^(FMDatabase *db) {
-        FMResultSet *rs = [db executeQuery:@"SELECT id, url, ssl, activeaccount, storage_occupied, storage, has_share_api_support, has_sharee_api_support, has_cookies_support, has_forbidden_characters_support, has_capabilities_support, instant_upload, background_instant_upload, path_instant_upload, only_wifi_instant_upload, timestamp_last_instant_upload, url_redirected, sorting_type FROM users WHERE id = ?", [NSNumber numberWithInteger:idUser]];
+        FMResultSet *rs = [db executeQuery:@"SELECT id, url, ssl, activeaccount, storage_occupied, storage, has_share_api_support, has_sharee_api_support, has_cookies_support, has_forbidden_characters_support, has_capabilities_support, image_instant_upload, video_instant_upload, background_instant_upload, path_instant_upload, only_wifi_instant_upload, timestamp_last_instant_upload, url_redirected, sorting_type FROM users WHERE id = ?", [NSNumber numberWithInteger:idUser]];
```
Comment (nasli) [inline Owncloud iOs Client/DataBase/Queries/ManageUsersDB.m]: Should use two timestamps for different options, videos and photos

**Comment by nasli** [comment]
### Bug

Steps:

1. Enable instant upload for photos
2. Close the app (kill)
3. Take some photos and videos
4. Open app and enabled instant videos too
5. Close the app take some videos and photos

Actual Behaviour:
Only latest photos and videos are uploaded since the time instant video was enabled

Expected Behaviour:
All photos are uploaded and only the videos after enabled instant videos

iPhone 6+ 9.3.2

**Comment by jesmrec** [comment]
Removing the QA Approved until bug in last message is solved.

**Comment by jesmrec** [comment]
@Jon-Schneider if you are storing only one timestamp for both kind of instant uploads, won't be confusing when both were enabled and after that, one of them is disabled?

The point is: i could not reproduce any scenario, bit as @nasli told, both kind of instant uploads should be managed separately.

What do you think?

**Comment by Jon-Schneider** [comment]
@nasli @jesmrec The reason I used one timestamp for both is because I wrote it to do one query per upload to make the uploads more efficient. I can split them into two queries when I have the time. It's not going to be much more expensive to do so. 

I can also easily fix that issue by removing the delay before processing Instant Uploads on open in AppDelegate. Then all the most recent items will be queued before the app finishes opening and a user can turn off either Instant Upload option won't cause this bug. I retained the delay when I refactored (I think it's 4 seconds) because the old version of Instant Upload had a delay.

Thoughts?

**Comment by jesmrec** [comment]
I think it is more secure to save both timestamps separately.

**Comment by nasli** [comment]
Thanks @Jon-Schneider , I will strongly recommend to retain the delay and use different timestamp for both, could be the case that too many photos pending and not time to process all before any change

**Comment by mmattel** [comment]
First of all, great, please proceed and get it merged and rolled out. Really needed !

Second. Please consider, maybe as a separate follow up PR, that there could be more accounts present. Currently it is uploading to the active account, but this maybe totally contraproductive as it maybe the wrong one. It would be great to have a setting available when multiple accounts are present, to chose the one which is used for uploading. Background: without, uploads would get totally fragmented and maybe get to the wrong account uploaded.

**Comment by jesmrec** [comment]
@mmattel you are right. We are developing the feature step by step, so after the current PR other features will come, like the one you mentioned .

**Comment by Jon-Schneider** [comment]
@jesmrec @nasli Finished.

**Comment by jesmrec** [comment]
@Jon-Schneider @nasli 

Instant uploads working as expected. 

The problem of conflicts persists (https://github.com/owncloud/ios/issues/755#issuecomment-247583486). Do you think it should be handled in another issue or in the current one?

**Comment by Jon-Schneider** [comment]
Could this be an issue with file uploads, not Instant Upload itself?

**Comment by jesmrec** [comment]
@Jon-Schneider i can not reproduce it with manual uploads. Only with instant uploads.

**Comment by Jon-Schneider** [comment]
@jesmrec I've had trouble reproducing it, which has made it hard to track down. I just pushed a commit that synchronizes the Instant Upload method. Did this help? I haven't been able to reproduce it after opening the app, taking photos, and reopening the app after half a dozen tries.

**Comment by jesmrec** [comment]
The problem persists. So, we should merge this PR and handle it in another issue.

I will open it, ping you, and attach all the info as possible.

Thanks for your contribution @Jon-Schneider

**Comment by nasli** [comment]
Sorry @Jon-Schneider , due to new changes in master branch a new conflict appears, to fix this conflict you can use filemerge tool (Xcode>Open Developer Tool>FileMerge) and compare this two file, probably the problem is just a change in one string

**Comment by Jon-Schneider** [comment]
@nasli Rebased.

@jesmrec I'll try to resolve #803 as quickly as possible.

**Comment by nasli** [comment]
Great!! Thanks @Jon-Schneider 👍

**Pull request merged**